2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,493 @@
+  if (power_before >= max_inactive.voting_power) then
+    //if active but it loses power below the max_inactive then move validator to inactive
+    //and promote max_inactive
+    if (power_after < max_inactive.voting_power) then
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, max_inactive)
+      remove(validator_sets[epoch].inactive, max_inactive.validator)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+    //if active and its change in power does not degrade it to inactive, then update its voting power
+    else
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+  else
+    //if inactive and gains power above the min_active, then insert into active and
+    //degrade min_active 
+    if (power_after > min_active.voting_power) then
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, min_active)
+      remove(validator_sets[epoch].active, min_active.validator)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+    //if inactive and its change in power does not promote it to active, then update its voting power
+    else
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+}
+```
+
+```go
+func remove_from_validator_sets(validator_address)
+{
+  var max_inactive = last(validator_sets[epoch].inactive)
+  remove(validator_sets[epoch].active, validator_address)
+  remove(validator_sets[epoch].inactive, max_inactive.validator)
+  add(validator_sets[epoch].active, max_inactive)    
+}
+```
+
+```go
+func is_validator(validator_address, epoch){
+    return read_epoched_field(validators[validator_address].state, epoch, ⊥) == candidate
+}
+```
+
+```go
+end_of_epoch()
+{
+  set_validators = {val | val = slash.validator && slash in enqueued_slashes[cur_epoch]}
+  forall (validator_address in set_validators) do
+    //jail validator (Step 1.2 of cubic slashing). This also unfreeze the validator (Step 2.5 of cubic slashing)
+    validators[validator_address].state[cur_epoch+1] = jailed
+    remove_from_validator_sets(validator_address, cur_epoch+1)
+    //iterate over all slashes for infractions within (-1,+1) epochs range (Step 2.1 of cubic slashing)
+    var set_slashes = {s | s in enqueued_slashes[epoch] && cur_epoch-1 <= epoch <= cur_epoch+1 && s.validator == validator_address}
+    //calculate the slash rate (Step 2.2 of cubic slashing)
+    var rate = compute_slash_rate(set_slashes)
+    forall (slash in {s | s in enqueued_slashes[cur_epoch] && slash.validator == validator_address}) do
+      //set the slash on the now "finalised" slash in storage (Step 2.3 of cubic slashing)
+      slash.rate = rate
+      append(slashes[cur_epoch], slash)
+      slashed_amount = slash.voting_power * slash.rate
```

Comment:
Should we _unfreeze_ validators here?
Otherwise, how does a frozen validator leave that state?

Reply:
Oh, I see they get jailed. Please ignore my comment